### PR TITLE
feat(unified-memory): read unified pool from attention backends fa3/flashinfer/trtllm_mha/flashmla

### PR DESCRIPTION
**Pull request by caihuali95**
> **Note on scope — this PR was split.** It previously contained both the read-path refactor 
> and this backend migration. The refactor is now a separate PR that this one stacks on, so 
> the mechanism can be reviewed independently of the per-backend work. 
> Stack order: **#34602 → #35245 → #35247 → this PR**.
>
> Only the last 8 commits are for this PR. The other commits are from the stacked PRs.

## Motivation

With the read path behind a single choke point, the remaining attention backends can read 
the unified memory pool without carrying any id-space logic of their own. This PR does that 
migration and widens the server's allow-list accordingly, which is what lifts the unified pool 
from Triton-only to the backends users actually run.

## Modifications

Migrates FlashAttention (fa3/fa4), FlashInfer's MHA and MLA backends, trtllm_mla together
with the two backends that subclass it, and trtllm_mha onto the choke point's canonical page
tables. Each backend keeps its existing builder for non-unified pools and dispatches to the 
shared one only when the source is active, so the change is inert off the unified path. The MLA 
model doors' read indices are translated at their production sites (the deepseek MHA mixin), 
which lets both pool doors become passthrough and deletes the last pool-resident translate 
hook. The now-unused backend hook shim is deleted and replaced by an enforcement scan, so 
the boundary cannot be re-crossed silently. The startup allow-list is widened to match what the backends 
can now serve — fa3/fa4/FlashInfer/trtllm_mha for the MHA and SWA families, plus flashmla 
for MLA — and end-to-end cells pin the resulting matrix.

| model class | before | after |
|---|---|---|
| unified MHA/SWA (uniform rows) | `triton` | `triton, fa3, fa4, flashinfer, trtllm_mha` |
| unified MLA | `triton, fa3, trtllm_mla, flashinfer, cutedsl_mla, tokenspeed_mla` | + `flashmla` |
| asymmetric K/V, or env-forced strided | `triton` | `triton` (unchanged) |

`unified_mem_hooks.py` is deleted. After the fold, two translate residents remain
outside the choke point, each by design and neither reachable from a backend: the
allocator-internal v2p implementations (the mechanism the choke point calls) and
the PD transfer-plane `translate_kv_indices_for_transfer` (RDMA staging, outside
the forward path). **A registered source scan enforces this** — it fails if any
file under `layers/attention/` calls a unified translate, probes for one, or
imports the deleted hooks module.

> **Note:** the accuracy and speed tables below were measured on the previous revision of this stack. The series has since been restructured per review and rebased onto current main; a revalidation pass is in progress and the tables will be refreshed.

## Accuracy Tests

GSM8K, unified vs. baseline **at matched backend**. Every newly enabled backend is at parity.

| Backend | Median Δ |
|---|---:|
| Triton | +0.00 pt |
| **fa3** | **+0.00 pt** |
| **FlashInfer** | **+0.00 pt** |
| **flashmla** | **+0.00 pt** |

## Speed Tests and Profiling

Serving benchmark at matched batch, production configuration: radix
cache + overlap scheduler + cuda graphs + page_size 256.

Two workloads:
heavy-decode (512-token prompts, 128 concurrent) and radix-retract
(4k shared-prefix prompts, 192 concurrent, forcing eviction and retraction).

Positive = unified is worse. Each row is heavy-decode / radix-retract.

| model | backend | ITL |
|---|---|---|
| Qwen3.5-9B (GDN) | triton | +0.55% / −0.29% |
| | fa3 | +0.10% / −0.20% |
| | flashinfer | +0.35% / −0.22% |
| gpt-oss-20b (SWA) | triton | −1.13% / −0.15% |
| | fa3 | +0.35% / +0.04% |
| Falcon-H1-7B (Mamba) | triton | +1.47% / −0.38% |

## Checklist

- [ ✅ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ✅ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ✅ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ✅ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #33366081537](https://github.com/sgl-project/sglang/actions/runs/33366081537)
Latest PR Test (Extra): :x: [Run #33366158884](https://github.com/sgl-project/sglang/actions/runs/33366158884)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33366081659](https://github.com/sgl-project/sglang/actions/runs/33366081659)

### CODE REVIEW

File: python/sglang/kernels/ops/kvcache/kv_indices.py [hunk 1]
```diff
@@ -7,19 +7,33 @@
...
     req_pool_indices_ptr,
     page_kernel_lens_ptr,
     kv_indptr,
     kv_start_idx,
     kv_indices_ptr,
-    req_to_token_ptr_stride: tl.constexpr,
+    # Runtime, not constexpr: the choke point's eager canonical is allocated at
+    # the batch's live width, so a constexpr stride would JIT-specialize per
+    # width (a recompile every few decode steps at small page sizes).
+    req_to_token_ptr_stride,
```
Comment (ch-wan) [inline python/sglang/kernels/ops/kvcache/kv_indices.py]: This is the right fix and the reasoning in the comment is exactly right. Worth flagging that **#35247 does not have it** — it adds `ENTRY_PAGE_SIZE` to this same signature while leaving `req_to_token_ptr_stride: tl.constexpr`, and then passes the per-batch `index_table.row_stride`. I have left a comment there pointing at this change; whichever of the two lands first, the other needs it.

File: python/sglang/kernels/ops/kvcache/kernel_page_table.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    n_pages = (seqlen + PAGE_SIZE - 1) // PAGE_SIZE
+
+    cols = blk * BLOCK + tl.arange(0, BLOCK)
+    mask = cols < n_pages
+    tok = tl.load(
+        req_to_token_ptr + req * req_stride + cols.to(tl.int64) * PAGE_SIZE,
+        mask=mask,
+        other=0,
+    ).to(tl.int64)
+    # Truncating int division routes negative (padding) tokens to page 0, whose
```
Comment (ch-wan) [inline python/sglang/kernels/ops/kvcache/kernel_page_table.py]: The reasoning here is right for `PAGE_SIZE > 1` and wrong for `PAGE_SIZE == 1`.

Truncating division does route a negative token to page 0 when `ps > 1`, and `MultiEndedAllocator.clear()` pins `virtual_to_physical[0] = 0`, so "page 0's v2p entry is the identity anchor" holds and the lane lands on the sink. Good — and worth naming that pin in the comment, since the argument silently depends on it.

At `PAGE_SIZE == 1` the premise fails: `-1 // 1 == -1`, so the next line loads `v2p_ptr[-1]`, one element before the table. Confirmed on an H200 against the sibling copy of this kernel in #35247 (identical arithmetic) by backing `v2p` with a tensor whose preceding element I controlled:

```
req_to_token[0, :4] = [0, 1, -1, 3],  v2p[i] = 101+i,  mult = 8
backing[0] = 123456789   # the int64 immediately before v2p[0]
kernel out : [808, 816, 987654312, 832]
                          ^ == 123456789 * 8
```

The garbage survives `tl.maximum(phys*mult, 0)` and enters the page table as a KV id far outside the pool. Note the CPU reference in `build_kernel_page_table` below already guards this (`torch.where(tok < 0, 0, tok // page_size)`), so the two implementations of this function disagree — and CI only exercises the CPU one.

I could not find a mainline path that puts `-1` inside a row's live prefix, so I do not believe it is live. But `tl.where(tok < 0, 0, tok // PAGE_SIZE)` costs one instruction, makes the two halves agree, and removes the `ps == 1` case from the comment's caveats.
